tablo/models.py

```diff
@@ -432,25 +451,32 @@ def _validate_where_clause(self, where):
         parsed = self._parse_where_clause(where)
 
         if parsed is None:
-            return True
+            return
         elif parsed[1]:
-            return False  # More than one statement means SQL injection.
-        else:
-            return self._validate_fields(parsed[0])
+            raise SQLInjectionError('Invalid where clause')
 
-    def _validate_fields(self, fields):
-        query_fields = {field.replace('"', '') for field in fields if '*' not in field}
+        self._validate_fields(parsed[0])
+
+    def _validate_fields(self, fields, include_related=True):
+        if isinstance(fields, str):
+            fields = fields.split(',')
+
+        query_fields = {field.replace('"', '') for field in fields if field != '*'}
         if not query_fields:
-            return True
-        else:
-            valid_fields = set(r for r in self.related_fields).union(f['name'] for f in self.fields)
-            return query_fields.issubset(valid_fields)
+            return
+
+        valid_fields = set(f['name'] for f in self.fields)
+        if include_related:
+            valid_fields = valid_fields.union('{0}.*'.format(r.related_title) for r in self.relations)
+            valid_fields = valid_fields.union(f for f in self.related_fields)
+
+        invalid_fields = query_fields.difference(valid_fields)
+        if invalid_fields:
+            raise InvalidFieldsError('Invalid fields: {0}'.format(', '.join(invalid_fields)))
```

Comment:
It's a good point. I wrestled with this a bit because this validation happens at the model level, and yet is validating based on user inputs that determine context.

It seemed better to me at the time to handle all field validation consistently. At this level, all we know is whether the `include_related` flag is `True` or `False`. We don't assume context, we just validate. If I try to add a contextual message here, then the input processing encapsulated in `perform_query` starts to bleed everywhere.

Still, one possible way to address this concern would be to raise a different error type if `include_related` is False and there is a related field in the list. I could call it something like `RelatedFieldsError`, which at this level could mean that the current query isn't going to do a join, or is intended only to return source table fields for `model`-like reasons. That way `perform_query` can catch the `RelatedFieldsError` it expects for its own reasons, and can send up a more informative error message to the view. Does this seem reasonable to you?